src/Transport/Receiving/RepeatedFailuresOverTimeCircuitBreaker.cs

```diff
@@ -23,9 +23,8 @@ public RepeatedFailuresOverTimeCircuitBreaker(string name, TimeSpan timeToWaitBe
 
         public void Success()
         {
-            var oldValue = Interlocked.Exchange(ref failureCount, 0);
-
-            if (oldValue == 0)
+            // If the failure count was already zero, replace it with zero (no change) and then return original
+            if (Interlocked.CompareExchange(ref failureCount, 0, 0) == 0)
```

Comment:
This seems like it would solve both problems.

//if there have been no failures previously then return
if(Interlocked.Read(ref failureCount) == 0 )
{
	return;
}
else //if there were previous failures then it's ok to disarm now.
{
   if(Interlocked.Read(ref failureCount) > 0) //not sure if this check is necessary or not.
   {
	  Interlocked.Exchange(ref failureCount, 0);
   }
}